### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by JennyVoPhamhi**
Edit June 23: Ready to merge Refactor into master

### CODE REVIEW

File: fishanno/clus.py
Comment (kevinyamauchi) [inline fishanno/clus.py]: I am not sure this is the preferred pattern for iterating over a pandas dataframe. See [`DataFrame.iterrows`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iterrows.html) and some notes in [this stackoverflow post](https://stackoverflow.com/questions/16476924/how-to-iterate-over-rows-in-a-dataframe-in-pandas).

I don't think this is critical to fix, but it would be nice if you have time. I mostly raised it so you know for the future.

File: fishanno/clus.py
Comment (kevinyamauchi) [inline fishanno/clus.py]: I think it would be nice to return at least the figure and maybe also the axes in case the user wants to modify the plot.

File: fishanno/clus.py
Comment (kevinyamauchi) [inline fishanno/clus.py]: See my note above about iterating over dataframes.

File: fishanno/clus.py
Comment (kevinyamauchi) [inline fishanno/clus.py]: While this totally works, you can iterate over lists and get the index like this:
```python
for i, small_cluster in enumerate(small_clusters_list):
    small_clusters['centroid_x'][i] = small_cluster[0]
```
See this [stackoverflow post](https://stackoverflow.com/questions/522563/accessing-the-index-in-for-loops) for more info. I think there are a few instances below with the same pattern.

File: fishanno/clus.py
Comment (kevinyamauchi) [inline fishanno/clus.py]: in general, all caps is reserved for constants ([PEP8](https://www.python.org/dev/peps/pep-0008/#constants))

**Review by kevinyamauchi — COMMENTED** [review]
Hey @JennyVoPhamhi . This looks good to me! I haven't actually ran the test notebook yet, but I have a few comments below. Once you update the installer, I will run the test notebook. Please let me know if you have any questions!

- I think we should have a `requirements.txt` file that lists the dependencies. You can then parse them in the `setup.py`. See the [packaging docs](https://packaging.python.org/discussions/install-requires-vs-requirements/) for some info on the requirements file and `install_requires`. I prefer to have the requirements file so that it is clear to the user what is required and they can do `pip install -r requirements.txt`. Additionally, parse the requirements file in the setup.py file and use `install_requires` so the user can use `pip install -e .`. See the [imagingDB](https://github.com/czbiohub/imagingDB) for an implementation example.
- I think we should add an installation section to the readme. I think something simple like what is in the [starfish docs](https://spacetx-starfish.readthedocs.io/en/latest/getting_started/installation/index.html) is sufficient.
- In general, it's preferred if the tests results are `True`/`False` so the results are easy to interpret.  For example, rather than comparing images, I think it would be better to compare the underlying coordinates to make sure they annotations are correctly identified. What do you think?
- Can you also use [`flake8`](http://flake8.pycqa.org/en/latest/) to lint the code?

File: fishanno/clus.py
Comment (kevinyamauchi) [reply fishanno/clus.py]: I think there are a few instances where this applies.

**Comment by JennyVoPhamhi** [comment]
Hi @kevinyamauchi, thank you for all this feedback. I'll get them implemented and let you know as soon as they're done. Thank you especially for the explanations and teaching points!

File: fishanno/clus.py
Comment (JennyVoPhamhi) [reply fishanno/clus.py]: Cool, fixed all instances of this! Pushing soon

File: fishanno/clus.py
Comment (JennyVoPhamhi) [reply fishanno/clus.py]: Awesome, this looks so much nicer!

File: fishanno/clus.py
Comment (JennyVoPhamhi) [reply fishanno/clus.py]: Cool, I've made this change!

**Comment by JennyVoPhamhi** [comment]
Hi @kevinyamauchi,

I'm working on changing the package name from fishanno to insta-pipeline and trying to make sure everything still works and runs. Sorry for the additional delay, but would you please hold off a little longer before changing the FISH-annotation repo name and doing the PR?

Thank you,

Jenny

**Comment by kevinyamauchi** [comment]
Thanks for letting me know! No problem!

**Comment by JennyVoPhamhi** [comment]
Thank you, Kevin!

**Comment by kevinyamauchi** [comment]
@JennyVoPhamhi , I just wanted to check in and see if this is ready for review. No worries if not!

**Comment by JennyVoPhamhi** [comment]
Thank you for checking! I'm really sorry, it's not ready yet. With
graduation next week, final stuff, and other wrap-up logistics, things have
been a bit crazy. I'm really looking forward to giving the repo focused
time, hopefully this weekend. I'll keep you posted!

**Comment by kevinyamauchi** [comment]
I just wanted to make a quick note that we might want to add jupyter notebooks as a dependency so people can just pip install and then run the examples.

**Comment by JennyVoPhamhi** [comment]
Sounds good to me. Will do!

File: README.md
Comment (kevinyamauchi) [inline README.md]: I don't think we should suggest that users install with "editable" mode (`-e`) since they will likely just be running the code.
```suggestion
- Install: `pip install <path to directory with setup.py>`
```

File: instapipeline/QuantiusAnnotation.py
Comment (kevinyamauchi) [inline instapipeline/QuantiusAnnotation.py]: Please update all imports to the new package name.

File: instapipeline/QuantiusAnnotation.py
Comment (kevinyamauchi) [inline instapipeline/QuantiusAnnotation.py:15]: Please document the input arguments in a docstring.

Comment (kevinyamauchi) [reply instapipeline/QuantiusAnnotation.py:15]: It looks like a couple of the other classes also need their `__init__` methods documented. Can you please go through them and update?

File: instapipeline/__init__.py
Comment (kevinyamauchi) [inline instapipeline/__init__.py]: Please update the name

File: instapipeline/autocrop.py
Comment (kevinyamauchi) [inline instapipeline/autocrop.py:39]: It looks like a lot of the functions don't have the parameters and returned variables documented in their docstrings. At a minimum, can you please go through and document all of the "public ones" (i.e., the ones you would actually call during your analysis). I think it is okay for now if you just describe what the function does for the more "utility" functions (i.e., the functions that aren't directly called by the user, but are called by the "public" functions).

**Review by kevinyamauchi — CHANGES_REQUESTED** [review]
Thanks, @JennyVoPhamhi ! I think this generally looks good. There are a couple of minor things that need to be addressed before we can merge:

- I wasn't able to run the demo notebooks because the import names need to updated to the new package name. 
- I noticed that there are quite a few public classes and methods/functions without docstrings. Can you please add them (see my comments below)?

**Comment by JennyVoPhamhi** [comment]
Sorry about that with the import statements! The demo notebooks did run for me and I think that’s because I already had fishanno installed. This time when I test before letting you know it’s ready, I’ll create a fresh new environment and run the demo notebooks.

Yes, I'll add all those doc strings.

File: instapipeline/QuantiusAnnotation.py
Comment (JennyVoPhamhi) [reply instapipeline/QuantiusAnnotation.py]: Sorry about this! Did this and checked the demo notebooks again.

File: instapipeline/__init__.py
Comment (JennyVoPhamhi) [reply instapipeline/__init__.py]: Did this. Thank you for catching!

File: instapipeline/autocrop.py
Comment (JennyVoPhamhi) [reply instapipeline/autocrop.py:39]: Got it. Will do!

**Comment by JennyVoPhamhi** [comment]
Hi @kevinyamauchi,

Thank you for your feedback and sorry again about the issue with the imports. I know you had specifically set aside time to do the PR over the weekend. The demo notebooks did run for me, even with “import fishanno,” and I think that is because I already had fishanno installed in the virtual env I used to test them.

I think the codebase is now ready for you to complete the PR.

- I fixed the import statements and searched all code to make sure the name “fishanno” everywhere has been changed to “instapipeline”.

- I fixed/wrote all the doc strings for the public methods (and all the relevant utility methods where I think a more detailed explanation of parameters and return values would be useful to the user).

- I went through and flake8’d all the code in the codebase to clean up after my edits.

- I created a fresh new virtual environment, installed instapipeline, and tested that the demo notebooks run.

Strangely, when jupyter notebook is included in the requirements.txt folder and installed with all the other requirements, I am not able to import instapipeline from a jupyter notebook, even though I am able to import from python in the terminal

$ python3
>> import instapipeline

So I removed `notebook==6.0.3` from requirements.txt and updated the installation instructions so that the user runs `conda install -c conda-forge notebook` inside the virtual environment. 

Do you think it’s okay to leave it like this?

Thank you,

Jenny

**Review by kevinyamauchi — APPROVED** [review]
Hey @JennyVoPhamhi ! Thank you for making the requested changes. I think the updated requirements + note in the readme is fine. I ran the demos with a fresh install and they looked good. Also, thank you for updating the docstrings. I have left a few minor, nitpicks below. I am approving and I think you can merge once you are happy with your final edits!

- I noticed some of the notebooks (e.g., the demos) have an empty cell at the end, can you remove the empty cells?
- In general, I think it's best to put the types in the descriptions of the parameters in the docstrings since it isn't easy to know in python from inspection. I see that you have it in some of them, so I would say just add the ones that you think are important or potentially confusing.

Thanks again! This is really great!

**Comment by JennyVoPhamhi** [comment]
Hi @kevinyamauchi ! Awesome, thank you very much for this feedback! Sounds great -- I will make those changes and merge, and then I will let you know that we are ready to change the repo name.

**Comment by JennyVoPhamhi** [comment]
Hi @kevinyamauchi ! I have…

- removed empty cells from all notebooks in FISH-annotation
- gone through all docstrings in codebase to check that the types are in the descriptions of the parameters, and more generally combed through for errors/typos
- flake8’ed everything in the codebase
- tested demo notebooks to make sure they run

So I am merging!

**Pull request merged**
